homeassistant/components/ccm15/coordinator.py

```diff
@@ -0,0 +1,332 @@
+"""Climate device for CCM15 coordinator."""
+import asyncio
+import datetime
+import logging
+from typing import Any, Optional
+
+import aiohttp
+import httpx
+import xmltodict
+
+from homeassistant.components.climate import (
+    FAN_AUTO,
+    FAN_HIGH,
+    FAN_LOW,
+    FAN_MEDIUM,
+    SWING_OFF,
+    SWING_ON,
+    ClimateEntity,
+    ClimateEntityFeature,
+    HVACMode,
+)
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    BASE_URL,
+    CONF_URL_CTRL,
+    CONF_URL_STATUS,
+    CONST_CMD_FAN_MAP,
+    CONST_CMD_STATE_MAP,
+    CONST_FAN_CMD_MAP,
+    CONST_STATE_CMD_MAP,
+    DEFAULT_TIMEOUT,
+    DOMAIN,
+)
+from .data_model import CCM15DeviceState, CCM15SlaveDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class CCM15Coordinator(DataUpdateCoordinator[CCM15DeviceState]):
+    """Class to coordinate multiple CCM15Climate devices."""
+
+    def __init__(
+        self, host: str, port: int, interval: int, hass: HomeAssistant
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=host,
+            update_method=self._async_update_data,
+            update_interval=datetime.timedelta(seconds=interval),
+        )
+        self._host = host
+        self._port = port
+        self._ac_devices: dict[int, CCM15Climate] = {}
+
+    def get_devices(self):
+        """Get all climate devices from the coordinator."""
+        return self._ac_devices.values()
+
+    async def _async_update_data(self) -> CCM15DeviceState:
+        """Fetch data from Rain Bird device."""
+        try:
+            return await self._fetch_data()
+        except httpx.RequestError as err:  # pragma: no cover
+            raise UpdateFailed(f"Error communicating with Device: {err}") from err
+
+    async def _fetch_xml_data(self) -> str:  # pragma: no cover
+        url = BASE_URL.format(self._host, self._port, CONF_URL_STATUS)
+        _LOGGER.debug("Querying url:'%s'", url)
+        async with httpx.AsyncClient() as client:
+            response = await client.get(url, timeout=DEFAULT_TIMEOUT)
+        return response.text
+
+    async def _fetch_data(self) -> CCM15DeviceState:
```

Comment:
> All API specific code has to be part of a third party library hosted on PyPi. Home Assistant should only interact with objects and not make direct calls to the API.

Please move the API code into a library. More information about it can be found [here](https://developers.home-assistant.io/docs/creating_component_code_review/#4-communication-with-devicesservices)

Reply:
Done